### PR TITLE
Add MAX32xxx counter timer driver

### PR DESCRIPTION
**Pull request by ozersa**
This PR adds MAX32xxx MCUs timer node and provide counter driver that based on timer peripheral.

Timer node and counter driver enabled for:
- MAX32655
- MAX32670
- MAX32672
- MAX32680
- MAX32690
- MAX32666
- MAX32662

tests\drivers\counter\counter_basic_api  output for some boards.

MAX32655EVKIT
image

MAX32690EVKIT
image

MAX32670EVKIT
image

MAX32672EVKIT
image

samples\drivers\counter\alarm output
image

Known issue (will be investigated after this PR):
LpTimer1 not pass test on MAX32655, MAX32690
LpTimer0/1 not pass test on MAX32672

### CODE REVIEW

**Comment by MaureenHelm** [comment]
Needs rebase

**Comment by ozersa** [comment]
> Needs rebase

Done.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by ozersa** [comment]
Rebased onto main to fix conflicts, 
Additionally [samples\drivers\counter\alarm](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/counter/alarm) enabled for MAX32 MCUs.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
@nordic-krch please take a look

**Comment by ozersa** [comment]
I would appreciate if anybody else have a chance to review it.

**Comment by MaureenHelm** [comment]
@ttmut please review

File: drivers/counter/counter_max32_timer.c [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+#define DT_DRV_COMPAT adi_max32_counter
+
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/clock_control/adi_max32_clock_control.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+
+#include <wrap_max32_tmr.h>
+
+/** MAX32 MCUs does not have multiple channel */
```
Comment (ttmut) [inline drivers/counter/counter_max32_timer.c:16]: Small typo
```suggestion
/** MAX32 MCUs does not have multiple channels */
```

File: drivers/counter/counter_max32_timer.c [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+	if (counter_cfg->ticks == 0) {
+		return -EINVAL;
+	}
+
+	if (counter_cfg->ticks != cfg->info.max_top_value) {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
```
Comment (ttmut) [inline drivers/counter/counter_max32_timer.c:83]: Nothing seems to be set in this function.

File: drivers/counter/counter_max32_timer.c [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+	const struct max32_tmr_config *cfg = dev->config;
+
+	return Wrap_MXC_TMR_GetPendingInt(cfg->regs);
+}
+
+static uint32_t api_get_top_value(const struct device *dev)
+{
+	const struct max32_tmr_config *cfg = dev->config;
+
+	return cfg->info.max_top_value;
```
Comment (ttmut) [inline drivers/counter/counter_max32_timer.c:96]: Is this value unmodifiable?

File: dts/bindings/timer/adi,max32-timer.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      are available:
+      - 0: "ADI_MAX32_PRPH_CLK_SRC_PCLK" Peripheral clock
+      - 1: "ADI_MAX32_PRPH_CLK_SRC_EXTCLK" External Clock
+      - 2: "ADI_MAX32_PRPH_CLK_SRC_IBRO" Internal Baud Rate Oscillator
+      - 3: "ADI_MAX32_PRPH_CLK_SRC_ERFO" External Radio Frequency Oscillator
+      - 4: "ADI_MAX32_PRPH_CLK_SRC_ERTCO" External Real-Time Clock Oscillator
+      - 5: "ADI_MAX32_PRPH_CLK_SRC_INRO" Internal Ring Oscillator
+      - 6: "ADI_MAX32_PRPH_CLK_SRC_ISO" Internal Secondary Oscillator
+      - 7: "ADI_MAX32_PRPH_CLK_SRC_IBRO_DIV8" IBRO/8
+      The target device might not support all option please take a look on
```
Comment (ttmut) [inline dts/bindings/timer/adi,max32-timer.yaml:31]:
```suggestion
      The target device might not support every option please take a look on
```

**Review by ttmut — CHANGES_REQUESTED** [review]

**Comment by ttmut** [comment]
Could also add overlay files for MAX32666 and MAX32675 in counter_basic_api test.

File: drivers/counter/counter_max32_timer.c (hunk 3 above)
Comment (ozersa) [reply drivers/counter/counter_max32_timer.c:96]: Not need to be modified, see https://github.com/analogdevicesinc/zephyr/blob/add-MAX32xxx-Counter-Timer/drivers/counter/counter_max32_timer.c#L321

**Comment by ttmut** [comment]
The number of commits can be greatly reduced by combining commits from a single author into a single commit.

**Review by ttmut — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**